### ISSUE TITLE
Improve the error message in `linalg.eigh` when using large matrices

### ISSUE THREAD

**Issue by gabitza-tech** [pytorch/pytorch#92141]
### 🐛 Describe the bug

I am using eigh() function from torch.linalg liberary on a [53000,53000] matrix.

```python
def eigDecompose(
    laplacian: torch.Tensor, cuda: bool, device: torch.device = torch.device('cpu')
) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Calculate eigenvalues and eigenvectors from the Laplacian matrix.
    """
    if cuda:
        if device is None:
            device = torch.cuda.current_device()
        laplacian = laplacian.float().to(device)
    else:
        laplacian = laplacian.float()
    lambdas, diffusion_map = eigh(laplacian)
    return lambdas, diffusion_map

```

I get the following error:
> Intel MKL ERROR: Parameter 8 was incorrect on entry to SSYEVD.
> [...]
> laplacian = laplacian.float()
>     lambdas, diffusion_map = eigh(laplacian)
>                              ~~~~ <--- HERE
>     return lambdas, diffusion_map
> RuntimeError: false INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1666642865319/work/aten/src/ATen/native/BatchLinearAlgebra.cpp":1462, please report a bug to PyTorch. linalg.eigh: Argument 8 has illegal value. Most certainly there is a bug in the implementation calling the backend library.

You can reproduce the error:
>>>import torch as t
>>>t.linalg.eigh(t.randn([53000,53000])).detach()
>>>
>>>Intel MKL ERROR: Parameter 8 was incorrect on entry to SSYEVD.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: false INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1666642865319/work/aten/src/ATen/native/BatchLinearAlgebra.cpp":1462, please report a bug to PyTorch. linalg.eigh: Argument 8 has illegal value. Most certainly there is a bug in the implementation calling the backend library.

Is there a way to split the matrix in smaller parts, calculate the eigh(), merge them and get the same result? From what I understood, the size is bigger than INT32 upper limit.

### Versions

PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=0, USE_CUDNN=OFF, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by janeyx99** [comment]
@lezcano From the linalg perspective, would it be a good idea to add boundchecking + return an error message from PyTorch?

**Comment by IvanYashchuk** [comment]
For `n=53000` the `lwork` argument of [`ssyevd`](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem/lapack-least-squares-eigenvalue-problem-driver/symmetric-eigenvalue-problems-lapack-driver/syevd.html) must be ≥ 2*n^2+ 6*n + 1, which a lot larger than 32bit-interface we are using can handle.
We should raise a better error message, this part is simple - introduce a check here:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebraKernel.cpp#L261

Supporting these large matrices is possible, but requires implementing https://github.com/pytorch/pytorch/issues/51720

**Comment by lezcano** [comment]
We should either document this or throw a better error, as this is an issue that keeps coming up over and over.

**Comment by gabitza-tech** [comment]
Thank you everybody for the clarifications!

**Comment by norabelrose** [comment]
I don't think this issue really should have been closed, because as @lezcano said the error message is very bad currently.

**Comment by lezcano** [comment]
We would accept a fix following https://github.com/pytorch/pytorch/issues/92141#issuecomment-1382241971
